src/Fusion/Plugin.hs

```diff
@@ -96,16 +99,198 @@ import Fusion.Plugin.Types
+    goLet path bndr@(NonRec _ expr1) = go (bndr : path) True expr1
+    goLet path (Rec bs) = bs >>= (\(b, expr1) -> goLet path $ NonRec b expr1)
+
+-------------------------------------------------------------------------------
+-- Core-to-core pass to mark interesting binders to be always inlined
+-------------------------------------------------------------------------------
+
+data ReportMode = ReportSilent | ReportWarn | ReportVerbose | ReportVerbose2
+
+getNonRecBinder :: CoreBind -> CoreBndr
+getNonRecBinder (NonRec b _) = b
+getNonRecBinder (Rec _) = error "markInline: expecting only nonrec binders"
+
+-- XXX we can possibly have a FUSE_DEBUG annotation to print verbose
+-- messages only for a given type.
+--
+-- XXX we mark certain functions (e.g. toStreamK) with a NOFUSION
+-- annotation so that we do not report them.
+
+showInfo
+    :: CoreBndr
+    -> DynFlags
+    -> ReportMode
+    -> Bool
+    -> [CoreBndr]
+    -> [([CoreBind], Alt CoreBndr)]
+    -> CoreM ()
+showInfo parent dflags reportMode failIt uniqBinders annotated =
+    when (uniqBinders /= []) $ do
+        -- XXX can we print the whole path leading up to this binder?
+        let showDetails (binds, c@(con,_,_)) =
+                let path = DL.intercalate "/"
+                        $ reverse
+                        $ map (showSDoc dflags . ppr)
+                        $ map getNonRecBinder binds
+                -- showSDoc dflags (ppr name) ++ ": " ++
+                in path ++ ": " ++
+                    case reportMode of
+                        ReportWarn -> showSDoc dflags (ppr con)
+                        ReportVerbose -> showSDoc dflags (ppr c)
+                        ReportVerbose2 ->
+                            showSDoc dflags (ppr $ head binds)
+                        _ -> error "transformBind: unreachable"
+        let msg = "In "
+                  ++ showSDoc dflags (ppr parent)
+                  ++ " binders "
+                  ++ showSDoc dflags (ppr uniqBinders)
+                  ++ " scrutinize data types annotated with "
+                  ++ showSDoc dflags (ppr Fuse)
+        case reportMode of
+            ReportSilent -> return ()
+            _ -> do
+                putMsgS msg
+                putMsgS $ DL.unlines $ map showDetails annotated
+        when failIt $ error "failing"
+
+markInline :: ReportMode -> Bool -> Bool -> ModGuts -> CoreM ModGuts
+markInline reportMode failIt transform guts = do
+    dflags <- getDynFlags
+    anns <- getAnnotations deserializeWithData guts
+    if (anyUFM (any (== Fuse)) anns)
+    then bindsOnlyPass (mapM (transformBind dflags anns)) guts
+    else return guts
+  where
+    transformBind :: DynFlags -> UniqFM [Fuse] -> CoreBind -> CoreM CoreBind
+    transformBind dflags anns (NonRec b expr) = do
+        let annotated = letBndrsThatAreCases (altsContainsAnn anns) expr
+        let uniqBinders = DL.nub (map (getNonRecBinder. head . fst) annotated)
```

Comment:
It is fine to pass recursive bindings to setInlineOnBndrs, currently it ignores recursive binders but in the future it'd replace them with a loop breaker